src/translate/pddl_parser/parsing_functions.py

```diff
@@ -42,49 +153,111 @@ def set_supertypes(type_list):
+        the_predicates = []
+        for no, entry in enumerate(alist):
+            with context.layer(f"Parsing {no}. predicate"):
+                if not isinstance(entry, list):
+                    context.error("Invalid predicate definition.",
+                                  syntax=SYNTAX_PREDICATE)
+                the_predicates.append(parse_predicate(context, entry))
+        return the_predicates
+
+
+def parse_function(context, alist, type_name):
+    with context.layer("Parsing function name"):
+        if not isinstance(alist, list) or len(alist) == 0:
+            context.error("Invalid definition of function.",
+                          syntax=SYNTAX_FUNCTION)
+        name = alist[0]
+        if not isinstance(name, str):
+            context.expected_word_error("Function name", name)
+    with context.layer(f"Parsing function '{name}'"):
+        arguments = parse_typed_list(context, alist[1:])
+        if not isinstance(type_name, str):
+            context.expected_word_error("Function type", type_name)
     return pddl.Function(name, arguments, type_name)
 
 
-def parse_condition(alist, type_dict, predicate_dict):
-    condition = parse_condition_aux(alist, False, type_dict, predicate_dict)
-    return condition.uniquify_variables({}).simplified()
+def parse_condition(context, alist, type_dict, predicate_dict):
+    with context.layer("Parsing condition"):
+        condition = parse_condition_aux(
+            context, alist, False, type_dict, predicate_dict)
+        return condition.uniquify_variables({}).simplified()
 
 
-def parse_condition_aux(alist, negated, type_dict, predicate_dict):
+def parse_condition_aux(context, alist, negated, type_dict, predicate_dict):
     """Parse a PDDL condition. The condition is translated into NNF on the fly."""
+    if not alist:
+        context.error("Expected a non-empty block as condition.")
     tag = alist[0]
     if tag in ("and", "or", "not", "imply"):
         args = alist[1:]
         if tag == "imply":
-            assert len(args) == 2
+            if len(args) != 2:
+                context.error("'imply' expects exactly two arguments.",
+                              syntax=SYNTAX_CONDITION_IMPLY)
         if tag == "not":
-            assert len(args) == 1
-            return parse_condition_aux(
-                args[0], not negated, type_dict, predicate_dict)
+            if len(args) != 1:
+                context.error("'not' expects exactly one argument.",
+                              syntax=SYNTAX_CONDITION_NOT)
+            negated = not negated
     elif tag in ("forall", "exists"):
-        parameters = parse_typed_list(alist[1])
-        args = alist[2:]
-        assert len(args) == 1
+        if len(alist) != 3:
+            context.error("'forall' and 'exists' expect exactly two arguments.",
+                          syntax=SYNTAX_CONDITION_FORALL_EXISTS)
+        if not isinstance(alist[1], list) or not alist[1]:
```

Comment:
I'm sure this works in python, but is there a reason why we don't first check for "existence" of the variable and only then whether or not it is a list.

Reply:
I can imagine that `not alist[1]` is there to prevent empty lists, so even if `alist[1]` is a list, we want to throw the error if that list is empty. Maybe it makes sense to split the error message, though. First check that `alist[1]` is indeed a list and complain about this if it is not the case, and then check that the list is not empty and complain differently if this doesn't hold. Or did I misunderstand something and a "non-empty block" can also be something different than a list?

Reply:
I would also read this as "if the element is no list or an empty list". The order of the tests makes sense to me: the first check is only passed if the object is a list (so for example, `None` wouldn't pass the first check). Then the second test checks if the list is empty. Flipping the checks would first cast the object to a boolean from whatever type it is and check if it is "truthy" before checking if that truthiness means what we expect (checking if list is empty.)

Reply:
In Python, if x is a collection, "if x" is the idiomatic way to test for non-emptiness. Splitting this into two tests with separate errors along the lines of "expected a list", "list cannot be empty" would make sense, and I agree they only make sense in this order.